linera-core/src/unit_tests/client_tests.rs

```diff
@@ -1643,76 +1680,1188 @@ where
+            epoch: Epoch::ZERO
+        }
+    );
+    assert_eq!(certificate.round, Round::SingleLeader(0));
+
+    let expected_round = Round::SingleLeader(1);
+    builder
+        .check_that_validators_are_in_round(chain_id, BlockHeight::from(1), expected_round, 3)
+        .await;
+
+    let round = loop {
+        let manager = client.chain_info().await.unwrap().manager;
+        if manager.leader == Some(Owner::from(pub_key1)) {
+            break manager.current_round;
+        }
+        clock.set(manager.round_timeout.unwrap());
+        assert!(client.request_leader_timeout().await.is_ok());
+    };
+    let round_number = match round {
+        Round::SingleLeader(round_number) => round_number,
+        round => panic!("Unexpected round {:?}", round),
+    };
+
+    // The other owner is leader now. Trying to submit a block should return `WaitForTimeout`.
+    let result = client
+        .transfer(None, Amount::ONE, Recipient::root(2), UserData::default())
+        .await
+        .unwrap();
+    let timeout = match result {
+        ClientOutcome::Committed(_) => panic!("Committed a block where we aren't the leader."),
+        ClientOutcome::WaitForTimeout(timeout) => timeout,
+    };
+    client.clear_pending_block();
+    assert!(client.request_leader_timeout().await.is_err());
+    clock.set(timeout.timestamp);
+    client.request_leader_timeout().await.unwrap();
+    let expected_round = Round::SingleLeader(round_number + 1);
+    builder
+        .check_that_validators_are_in_round(chain_id, BlockHeight::from(1), expected_round, 3)
+        .await;
+
+    loop {
+        let manager = client.chain_info().await.unwrap().manager;
+        if manager.leader == Some(Owner::from(pub_key0)) {
+            break;
+        }
+        clock.set(manager.round_timeout.unwrap());
+        assert!(client.request_leader_timeout().await.is_ok());
+    }
+
+    // Now we are the leader, and the transfer should succeed.
+    let _certificate = client
+        .transfer(None, Amount::ONE, Recipient::root(2), UserData::default())
+        .await
+        .unwrap()
+        .unwrap();
+    assert_eq!(
+        client.local_balance().await.unwrap(),
+        Amount::from_tokens(2)
+    );
+
+    let expected_round = Round::SingleLeader(0);
+    builder
+        .check_that_validators_are_in_round(chain_id, BlockHeight::from(2), expected_round, 3)
+        .await;
+
+    Ok(())
+}
+
+#[test_case(MemoryStorageBuilder::default(); "memory")]
+#[cfg_attr(feature = "storage-service", test_case(ServiceStorageBuilder::new().await; "storage_service"))]
+#[cfg_attr(feature = "rocksdb", test_case(RocksDbStorageBuilder::new().await; "rocks_db"))]
+#[cfg_attr(feature = "dynamodb", test_case(DynamoDbStorageBuilder::default(); "dynamo_db"))]
+#[cfg_attr(feature = "scylladb", test_case(ScyllaDbStorageBuilder::default(); "scylla_db"))]
+#[test_log::test(tokio::test)]
+async fn test_finalize_validated<B>(storage_builder: B) -> anyhow::Result<()>
```

Comment:
(This probably overlaps with `blob_test1` now. I wonder whether the two can be merged. Maybe in a later PR, of course.)

Reply:
https://github.com/linera-io/linera-protocol/issues/2304